---
title: "Twisted beech mutations"
author: "Sylvain Schmitt"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  bookdown::html_document2:
    number_sections: false  
    toc: true
    toc_float: yes
  bookdown::markdown_document2:
    number_sections: false  
    toc: true
linestretch: 1.5
---

```{r setup, include=FALSE}
# libraries
set.seed(42)
library(knitr)
library(tidyverse)
library(vroom)
# bayesplot
library(dtplyr)
library(circlize)
theme_set(bayesplot::theme_default())
opts_chunk$set(
  echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
  cache = F, cache.lazy = F)
```

```{r snakemake, include=FALSE}
genomes_indexes <- snakemake@input[["index"]]
coverages_files <- snakemake@input[["coverage"]]
circos_cov <- snakemake@input[["circos_cov"]]
mutations <- vroom(snakemake@input[["mutations"]]) %>% 
  mutate(Filter = recode(Filter, "base" = "Base", "robust" = "EVS")) %>% 
  mutate(reference = recode(reference, "Fagus_sylvatica_mutant_v0" = "Mutant", 
                         "Fagus_sylvatica_revertant_v0" = "Revertant", "Fagus_sylvatica_v3" = "Ciron")) 
cv <- vroom(snakemake@input[["cv"]])
spectra <- vroom(snakemake@input[["spectra"]])
genomes <- lapply(genomes_indexes, function(i)
       read_tsv(i, col_names = c("name", "length", "bytesindex", "basesperline", "bytesperline")) %>% 
         mutate(file = i)) %>% 
  bind_rows() %>% 
  mutate(genome = gsub("results/reference/", "", file)) %>% 
  mutate(genome = gsub(".fa.fai", "", genome)) %>% 
  dplyr::select(genome, name, length) %>% 
  mutate(start = 0, end = length) %>% 
  dplyr::select(-length) %>%  
  mutate(trsc = letters[1:nrow(.)], exon = LETTERS[1:nrow(.)]) %>% 
  mutate(genome = recode(genome, "Fagus_sylvatica_mutant_v0" = "Mutant", 
                         "Fagus_sylvatica_revertant_v0" = "Revertant", "Fagus_sylvatica_v3" = "Ciron"))
coverages <- lapply(coverages_files, function(i)
  vroom::vroom(i, col_names = c("chrom", "start", "stop", "coverage")) %>% 
         mutate(file = i)) %>% 
  bind_rows() %>% 
   mutate(library = gsub("results/alns/", "", file)) %>% 
  mutate(library = gsub(".regions.bed", "", library)) %>% 
  separate(library, c("library", "genome"), "_on_") %>% 
  mutate(genome = recode(genome, "Fagus_sylvatica_mutant_v0" = "Mutant", 
                         "Fagus_sylvatica_revertant_v0" = "Revertant", "Fagus_sylvatica_v3" = "Ciron")) 
```

# Introduction

A particular form of the European beech (Fagus sylvatica L) was described as ’Tortillard’ (var tortuosa Pépin) by Pépin (1861).
Here we propose to study the mutations accumulated between the base of the trunk of a twisted beech and a reverting branch. 
We do not expect to find a specific point mutation responsible for the reversion of the dwarf phenotype,
but to gain insight into the somatic mutation landscape in the dwarf tree.

# Material and methods

We analyzed two illumina libraries from two sectors (mutant and revertant) of accession #354 sampled on March 31st 2021 in the Verzy forest
Libraries were aligned against three references: (1) the genome from a beech from the Ciron, (2) the genome of the mutant sector, and (3) the genome of the revertant sector.
We used the *detectMutations* pipeline developed by Schmitt *et al.* (2021) to detect somatic mutations from mapped sequencing reads on a each genome reference.
Pair-end sequencing reads of every library are quality checked using *FastQC* before trimming using *Trimmomatic* (Bolger *et al.*, 2014)
keeping only paired-end reads without adaptors and a phred score above 15 in a sliding window of 4 bases.
Reads are aligned against the reference per chromosome using *BWA mem* with the option to mark shorter splits (Li & Durbin, 2009).
Alignments are then compressed using *Samtools view* in CRAM format, sorted by coordinates using *Samtools sort*, and indexed using *Samtools index* (Li *et al.*, 2009). 
Duplicated reads in alignments are marked using *GATK MarkDuplicates* (Auwera et al., 2013). 
Finally, the workflow uses *Strelka2* to detect mutations (Kim *et al.*, 2018).
We compared the pair of sample points with the revertant sector considered as the mutated sample and the mutant sector as the normal sample to distinguish mutations among branches from heterozygous sites and sequencing errors.
We kept candidate mutations with (1) a read depth for both the normal and mutated samples between half and two times the mean sequencing depth, (2) an absence of the mutated allele in the normal sample, and (3) a minimum of 5 copies of the mutated allele in the mutated sample.
In addition, we took advantage of the Empirical Somatic Score (ESS) calculated by *Strelka2* (Kim *et al.*, 2018) to build a conservative dataset of candidate mutations.
Finally, we cross-validated detected mutations across genomes by realigning 1-kb windows containing mutations from one genome to the other using *BLAT* (Kent, 2002) and each genome pairs.
We assessed for each mutations its class (transition or transversion), type (*e.g.* C->T) and spectra (genomic context, *e.g.* AAT),
and we assessed the cummulated number of mutations with increasing allelic frequency to.
All analyses are currently done in the the [`hetre` branch of the `detectMutations` repository](https://github.com/sylvainschmitt/detectMutations/tree/hetre).

# Results

## Genome

Analyzed genomes showed a good contiguity with a ew long scaffolds (Fig. \@ref(fig:genomeFig)) with a mean coverage between 52X and 81X with higher coverages in the mutant library and on the genome from the accession #354 (Fig. \@ref(fig:coverageFig)).

```{r genomeFig, fig.cap="Fagus sylvatica scaffolds ordered by size."}
ggplot(genomes, aes(reorder(name, end), end/10^6, fill = end > 10*10^6)) +
  geom_col() +
  coord_flip() +
  ylab("Length (mb)") + 
  xlab("Fagus sylvatica scaffolds ordered by size") +
  scale_fill_discrete(">10Mb") +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
  facet_wrap(~ genome, scales = "free")
```

```{r coverageFig, fig.cap="Mean coverage on a 10kb window"}
ggplot(coverages,  aes(coverage)) +
  geom_histogram(fill = "lightgrey") +
  xlim(0, 200) +
  facet_grid(genome ~ library, labeller = "label_both") +
  xlab("Mean coverage on a 10kb window") +
  geom_vline(aes(xintercept = coverage), col = "red", size = 1.2,
             data = group_by(coverages, genome, library) %>% 
               summarise(coverage = median(coverage))) +
  geom_text(aes(label = paste0(round(coverage), "X"),
                x = coverage - 15), y = 10000, col = "red",
            data = group_by(coverages, genome, library) %>% 
              summarise(coverage = median(coverage))) 
```

## Mutations

We found 33k to 36k mutations with base filtering across genomes, 
against  2.3k to 2.6k when using conservative filtering based on the empirical variant score (EVS) from *Strelka2* (Tab \@ref(tab:muttab)).
Most mutations showed low-allelic frequencies (< 0.15) with an higher distribution when using conservative filtering (Fig. \@ref(fig:mutfig)).

```{r muttab}
mutations %>% 
  group_by(reference, Filter) %>% 
  summarise(N = n()) %>% 
  reshape2::dcast(reference ~ Filter) %>% 
  kable(caption = "Number of mutations per filter and reference", 
        format.args = list(big.mark = " "))
```

```{r mutfig, fig.cap="Alleles-frequencies of filtered mutations."}
ggplot(mutations, aes(mutation_AF, col = Filter), fill = NA) +
  geom_density(size = 1.3) +
  xlab("Allelic frequency")  +
  facet_wrap(~ reference, nrow = 2) +
  theme(legend.position = c(0.9, 0.1))
```

```{r muttaball}
filter(mutations, reference == "Revertant", Filter == "EVS") %>% 
  dplyr::select(CHROM, POS, REF, ALT, mutation_altCountT1, mutation_refCountT1, normal_altCountT1, normal_refCountT1, mutation_AF) %>% 
  DT::datatable(colnames = c("scf", "pos", "ref", "alt", "rev alt", "rev ref", "mut alt", "mut ref", "AF"),
                caption = "Mutation position on the Mutant's genome using conservative filtering based on the empirical variant score (EVS).")
```

## Cross-validation

98.7% of mutations were detected across all genomes, with only 13 not found on the Ciron's genome and 5 not found on the Mutant's genome (Tab. \@ref(tab:cv)).

```{r cv}
cv %>%
  reshape2::dcast(SNV ~ reference) %>%
  mutate_all(funs(ifelse(is.na(.), 0, .))) %>%
  mutate(crossvalidation = "all") %>% 
  mutate(crossvalidation = ifelse(Fagus_sylvatica_revertant_v0  != 0 & Fagus_sylvatica_v3 == 0,
                                  "Mutant & Revertant only", crossvalidation)) %>% 
  mutate(crossvalidation = ifelse(Fagus_sylvatica_revertant_v0  == 0 & Fagus_sylvatica_v3 != 0,
                                  "Mutant & Ciron only", crossvalidation)) %>% 
  group_by(crossvalidation) %>% 
  summarise(N = n()) %>% 
  kable(caption = "Number of mutations by cross-validation succes.", col.names = c("Cross-validation", "N"),
        format.args = list(big.mark = " "))
```

## Spectra

Mutations show classic spectra with an enrichment in transversion (C->T and T->C, Fig. \@ref(fig:type)) and no specific enrichment in genomic context (Fig. \@ref(fig:spectra)).

```{r type, fig.cap="Context-dependent mutation depending on mutation types. Mutation types have been summarised into six main classes with blue columns for transversion compared to transition in red."}
spectra %>%
  ggplot(aes(type, fill = class)) +
  geom_bar() +
  xlab("Mutation type")
```

```{r spectra, fig.cap="Context-dependent mutation spectra. Mutation types have been summarised into six main classes, and then differentiated depending on their 5’ and 3’ genomic contexts."}
spectra %>%
  ggplot(aes(spectra)) +
  geom_bar() +
  facet_wrap(~ type, scales = "free_x") +
  theme(axis.text.x = element_text(angle = 90), axis.title.x = element_blank()) +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(size = 0.25, color = "black", fill = NA))
```

## Rate

The revertant sector shows an accumulation of $10^1$ to $10^4$ mutation from the mutant sector (Fig. \@ref(fig:rate)).

```{r rate, fig.cap="Cummulated number of mutations with increasing allelic frequency on the Revertant's genome."}
mutations %>% 
  filter(reference == "Revertant") %>% 
  arrange(desc(mutation_AF)) %>% 
  mutate(mutations = 1) %>% 
  mutate(cummulated_mutations = cumsum(mutations)) %>% 
  ggplot(aes(mutation_AF, cummulated_mutations)) +
  geom_line() +
  geom_point() +
  scale_y_log10() +
  xlab("Allelic frequency") + ylab("Cummulated number of mutations")
```

# Conclusion

The dwarf tree show a very high number of low-frequency mutations (Fig. \@ref(fig:rate)) with a non-negligible number of high frequency mutations (399 with AF>0.2) that have been validated with cross-validation across genomes.
