---
title: "Mutations Hetre"
author: "Sylvain Schmitt"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  bookdown::html_document2:
    number_sections: false  
    toc: true
    toc_float: yes
linestretch: 1.5
---

```{r setup, include=FALSE}
# libraries
set.seed(42)
library(knitr)
library(kableExtra)
options(knitr.table.format = "html") 
library(tidyverse)
# vroom
# bayesplot
library(dtplyr)
library(circlize)
theme_set(bayesplot::theme_default())
opts_chunk$set(
  echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
  cache = T, cache.lazy = F)
# snakemake
genomes_indexes <- snakemake@input[["index"]]
coverages_files <- snakemake@input[["coverage"]]
circos_cov <- snakemake@input[["circos_cov"]]
mutations <- read_tsv(snakemake@input[["mutations"]])
```

The analyses of mutations from the ["Faux de Verzy"](https://en.wikipedia.org/wiki/Faux_de_Verzy) are
currently done in the the [`hetre` branch of the `detectMutations` repository](https://github.com/sylvainschmitt/detectMutations/tree/hetre).

## Genome

```{r genomeData}
genomes <- lapply(genomes_indexes, function(i)
       read_tsv(i, col_names = c("name", "length", "bytesindex", "basesperline", "bytesperline")) %>% 
         mutate(file = i)) %>% 
  bind_rows() %>% 
  mutate(genome = gsub("results/reference/", "", file)) %>% 
  mutate(genome = gsub(".fa.fai", "", genome)) %>% 
  dplyr::select(genome, name, length) %>% 
  mutate(start = 0, end = length) %>% 
  dplyr::select(-length) %>%  
  mutate(trsc = letters[1:nrow(.)], exon = LETTERS[1:nrow(.)])
coverages <- lapply(coverages_files, function(i)
  vroom::vroom(i, col_names = c("chrom", "start", "stop", "coverage")) %>% 
         mutate(file = i)) %>% 
  bind_rows() %>% 
   mutate(library = gsub("results/alns/", "", file)) %>% 
  mutate(library = gsub(".regions.bed", "", library)) %>% 
  separate(library, c("library", "genome"), "_on_")
```

```{r genomeFig, fig.cap="Fagus sylvatica scaffolds ordered by size."}
ggplot(genomes, aes(reorder(name, end), end/10^6, fill = end > 10*10^6)) +
  geom_col() +
  coord_flip() +
  ylab("Length (mb)") + 
  xlab("Fagus sylvatica scaffolds ordered by size") +
  scale_fill_discrete(">10Mb") +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
  facet_wrap(~ genome, scales = "free")
```

```{r coverageFig, fig.cap="Mean coverage on a 10kb window"}
ggplot(coverages, aes(coverage)) +
  geom_histogram(fill = "lightgrey") +
  xlim(0, 200) +
  facet_grid(genome ~ library) +
  xlab("Mean coverage on a 10kb window") +
  geom_vline(aes(xintercept = coverage), col = "red", size = 1.2,
             data = group_by(coverages, genome, library) %>% 
               summarise(coverage = median(coverage))) +
  geom_text(aes(label = paste0(round(coverage), "X"),
                x = coverage - 15), y = 10000, col = "red",
             data = group_by(coverages, genome, library) %>% 
               summarise(coverage = median(coverage))) 
```

```{r circos1, fig.cap="Genome coverage on a 10-kb windows for mutant and revertant."}
# include_graphics(circos_cov)
```

## Mutations

We used the following filters:

* **Base**
   * The mutation is detected in the revertant
   * A null number of alternate allele count in the normal sample (`normal_altCountT1 == 0`)
   * A read depth for the two sample between half and two times the mean coverage (`normal_DP <= 150, normal_DP >= 25.5, mutation_DP <= 150, mutation_DP >= 25.5`)
   * A minimum of 5 alternate allele count in the mutated sample (`mutation_altCountT1 >= 5`), corresponding to a minimum allele frequency of 0.1 for a mean coverage of 50X
   * A maximum allelic frequency of 0.5 (`mutation_AF <= 0.5`)
* **Robust**
   * All filters from base
   * `Strelka 2` automatic filtering (`Filter == "PASS"`)

```{r muttab}
mutations %>% 
  group_by(reference, Filter) %>% 
  summarise(N = n()) %>% 
  reshape2::dcast(reference ~ Filter) %>% 
  kable(cpation = "Number of mutations per filter and reference", 
        format.args = list(big.mark = " "))
```

```{r mutfig, fig.cap="Alleles-frequencies of filtered mutations in hetre."}
ggplot(mutations, aes(mutation_AF, col = Filter), fill = NA) +
  geom_density(size = 1.3) +
  xlab("Allelic frequency")  +
  facet_wrap(~ reference, nrow = 2) +
  theme(legend.position = c(0.9, 0.1))
```

```{r mutmutantfig, fig.cap="Mutation position on the Mutant's genome (scaffolds > 5-Mb) for robust candidates."}
genome <- filter(genomes, genome == "Fagus_sylvatica_mutant_v0", end > 5*10^6)
ggplot(genome, aes(x = name, xend = name)) +
  geom_segment(aes(y = start, yend = end/10^6), size = 3, col = "lightgrey") +
  geom_point(aes(y = POS/10^6, col = mutation_AF),
             data = filter(mutations, reference == "Fagus_sylvatica_mutant_v0", Filter == "robust") %>%
               mutate(name = CHROM) %>%
               filter(name %in% genome$name)) +
  coord_flip() +
  ylab("Position (Mb)") +
  theme(axis.line.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank()) +
  viridis::scale_color_viridis("AF")
```

```{r}
filter(mutations, reference == "Fagus_sylvatica_mutant_v0", Filter == "robust") %>% 
  dplyr::select(CHROM, POS, REF, ALT, mutation_altCountT1, mutation_refCountT1, normal_altCountT1, normal_refCountT1, mutation_AF) %>% 
  DT::datatable(colnames = c("scf", "pos", "ref", "alt", "rev alt", "rev ref", "mut alt", "mut ref", "AF"),
                caption = "Mutation position on the Mutant's genome for robust candidates.")
```

# Cross-validation

We further filtered mutations when the position on the Ciron's and the Revertant's genomes corresponded to the Mutant's genome.

```{r}
# read_tsv("../results/mutations/cross_validation.tsv") %>% 
#   reshape2::dcast(SNV ~ reference) %>% 
#   mutate_all(funs(ifelse(is.na(.), 0, .))) %>% 
#   filter(Fagus_sylvatica_mutant_v0 == 0 | Fagus_sylvatica_revertant_v0 == 0 | Fagus_sylvatica_v3 == 0) %>% 
#   View()
```


