---
title: "Mutations Hetre"
author: "Sylvain Schmitt"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  bookdown::html_document2:
    number_sections: false  
    toc: true
    toc_float: yes
linestretch: 1.5
---

```{r setup, include=FALSE}
# libraries
set.seed(42)
library(knitr)
library(kableExtra)
options(knitr.table.format = "html") 
library(tidyverse)
# vroom
# bayesplot
library(dtplyr)
library(circlize)
theme_set(bayesplot::theme_default())
opts_chunk$set(
  echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
  cache = T, cache.lazy = F)
# snakemake
genomes_indexes <- snakemake@input[["index"]]
coverages_files <- snakemake@input[["coverage"]]
circos_cov <- snakemake@input[["circos_cov"]]
base_file <- snakemake@input[["base"]]
robust_file <- snakemake@input[["robust"]]
```

The analyses of mutations from the ["Faux de Verzy"](https://en.wikipedia.org/wiki/Faux_de_Verzy) are
currently done in the the [`hetre` branch of the `detectMutations` repository](https://github.com/sylvainschmitt/detectMutations/tree/hetre).

## Genome

```{r genomeData}
genomes <- lapply(genomes_indexes, function(i)
       read_tsv(i, col_names = c("name", "length", "bytesindex", "basesperline", "bytesperline")) %>% 
         mutate(file = i)) %>% 
  bind_rows() %>% 
  mutate(genome = gsub("results/reference/", "", file)) %>% 
  mutate(genome = gsub(".fa.fai", "", genome)) %>% 
  dplyr::select(genome, name, length) %>% 
  mutate(start = 0, end = length) %>% 
  dplyr::select(-length) %>%  
  mutate(trsc = letters[1:nrow(.)], exon = LETTERS[1:nrow(.)])
coverages <- lapply(coverages_files, function(i)
  vroom::vroom(i, col_names = c("chrom", "start", "stop", "coverage")) %>% 
         mutate(file = i)) %>% 
  bind_rows() %>% 
   mutate(library = gsub("results/alns/", "", file)) %>% 
  mutate(library = gsub(".regions.bed", "", library)) %>% 
  separate(library, c("library", "genome"), "_on_")
```

```{r genomeFig, fig.cap="Fagus sylvatica scaffolds ordered by size."}
ggplot(genomes, aes(reorder(name, end), end/10^6, fill = end > 10*10^6)) +
  geom_col() +
  coord_flip() +
  ylab("Length (mb)") + 
  xlab("Fagus sylvatica scaffolds ordered by size") +
  scale_fill_discrete(">10Mb") +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank()) +
  facet_wrap(~ genome, scales = "free")
```

```{r coverageFig, fig.cap="Mean coverage on a 10kb window"}
ggplot(coverages, aes(coverage)) +
  geom_histogram(fill = "lightgrey") +
  xlim(0, 200) +
  facet_grid(genome ~ library) +
  xlab("Mean coverage on a 10kb window") +
  geom_vline(aes(xintercept = coverage), col = "red", size = 1.2,
             data = group_by(coverages, genome, library) %>% 
               summarise(coverage = median(coverage))) +
  geom_text(aes(label = paste0(round(coverage), "X"),
                x = coverage - 15), y = 10000, col = "red",
             data = group_by(coverages, genome, library) %>% 
               summarise(coverage = median(coverage))) 
```

```{r circos1, fig.cap="Genome coverage on a 10-kb windows for mutant and revertant."}
# include_graphics(circos_cov)
```

## Mutations

We used the following filters:

* **Base**
   * The mutation is detected in both directions of the comparison mutant versus revertant
   * A null number of alternate allele count in the normal sample (`normal_altCountT1 == 0`)
   * All filters from raw
   * A read depth for the two sample between half and two times the mean coverage (`normal_DP <= 150, normal_DP >= 25.5, mutation_DP <= 150, mutation_DP >= 25.5`)
   * A minimum of 5 alternate allele count in the mutated sample (`mutation_altCountT1 >= 5`), corresponding to a minimum allele frequency of 0.1 for a mean coverage of 50X
* **Robust**
   * All filters from base
   * `Strelka 2` automatic filtering (`Filter == "PASS"`)

```{r muttab}
list(base = base_file, robust = robust_file) %>% 
  lapply(read_tsv) %>% 
  bind_rows(.id = "filter") %>% 
  group_by(filter, reference) %>% 
  summarise(N = n()) %>% 
  reshape2::dcast(reference ~ filter) %>% 
  kable(cpation = "Number of mutations per filter and reference", 
        format.args = list(big.mark = " "))
```

```{r mutfig, fig.cap="Alleles-frequencies of filtered mutations in hetre."}
list(base = base_file, robust = robust_file) %>% 
  lapply(read_tsv) %>% 
  bind_rows(.id = "filter") %>% 
  ggplot(aes(mutation_AF, col = filter), fill = NA) +
  geom_density(size = 1.3) +
  scale_color_discrete("Filter") +
  xlab("Allelic frequency")  +
  xlim(0.05, 0.5) +
  facet_wrap(~ reference, nrow = 2) +
  theme(legend.position = c(0.9, 0.1))
```

```{r mutmutantfig, fig.cap="Mutation position on the Mutant's genome (scaffolds > 5-Mb) for base (blue) and robust (red) candidates."}
genome <- filter(genomes, genome == "Fagus_sylvatica_mutant_v0", end > 5*10^6)
base <- read_tsv(base_file) %>% 
  filter(reference == "Mutant") %>% 
  mutate(name = CHROM) %>% 
  filter(name %in% genome$name)
robust <- read_tsv(robust_file) %>% 
  filter(reference == "Mutant") %>% 
  mutate(name = CHROM) %>% 
  filter(name %in% genome$name)
ggplot(genome, aes(x = name, xend = name)) +
  geom_segment(aes(y = start, yend = end/10^6), size = 3, col = "lightgrey") +
  geom_point(aes(y = POS/10^6), col = "blue", base) +
  geom_point(aes(y = POS/10^6), col = "red", robust) +
  coord_flip() +
  ylab("Position (Mb)") +
  theme(axis.line.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank())
```

# Cross-validation

We further filtered mutations when the position on the Ciron's and the Revertant's genomes corresponded to the Mutant's genome.



