---
title: "detect Mutations"
author: Sylvain Schmitt
date: April 20, 2021
output:
  github_document: 
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(kableExtra)
if(knitr:::is_html_output()) options(knitr.table.format = "html") 
if(knitr:::is_latex_output()) options(knitr.table.format = "latex") 
library(tidyverse)
library(vcfR)
theme_set(bayesplot::theme_default())
opts_chunk$set(echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
               cache = F, cache.lazy = F)
```

[`singularity` & `snakemake`](https://github.com/sylvainschmitt/snakemake_singularity) workflow to detect mutations with several alignment and mutation detection tools.

```{r dag, fig.cap="Direct acyclic graph."}
knitr::include_graphics("dag/dag.svg")
```

# Installation

- [x] Python ≥3.5
- [x] Snakemake ≥5.24.1
- [x] Golang ≥1.15.2
- [x] Singularity ≥3.7.3
- [x] This workflow

```{bash, eval=F, echo=T}
# Python
sudo apt-get install python3.5
# Snakemake
sudo apt install snakemake`
# Golang
export VERSION=1.15.8 OS=linux ARCH=amd64  # change this as you need
wget -O /tmp/go${VERSION}.${OS}-${ARCH}.tar.gz https://dl.google.com/go/go${VERSION}.${OS}-${ARCH}.tar.gz && \
sudo tar -C /usr/local -xzf /tmp/go${VERSION}.${OS}-${ARCH}.tar.gz
echo 'export GOPATH=${HOME}/go' >> ~/.bashrc && \
echo 'export PATH=/usr/local/go/bin:${PATH}:${GOPATH}/bin' >> ~/.bashrc && \
source ~/.bashrc
# Singularity
mkdir -p ${GOPATH}/src/github.com/sylabs && \
  cd ${GOPATH}/src/github.com/sylabs && \
  git clone https://github.com/sylabs/singularity.git && \
  cd singularity
git checkout v3.7.3
cd ${GOPATH}/src/github.com/sylabs/singularity && \
  ./mconfig && \
  cd ./builddir && \
  make && \
  sudo make install
# detect Mutations
git clone git@github.com:sylvainschmitt/detectMutations.git
cd detectMutations
```

# Usage

## Get data

*Generate data using the [generate Mutations](https://github.com/sylvainschmitt/generateMutations) workflow.*

```{bash, eval=F, echo=T}
git clone git@github.com:sylvainschmitt/generateMutations.git
cd ../generateMutations
snakemake --use-singularity --cores 4
cd ../detectMutations
bash scripts/get_data.sh
```

## Locally

```{bash, eval=F, echo=T}
snakemake -np # dry run
snakemake --dag | dot -Tsvg > dag/dag.svg # dag
snakemake --use-singularity --cores 4 # run
snakemake --use-singularity --cores 1 --verbose # debug
snakemake --report report.html # report
```

## HPC

```{bash, eval=F, echo=T}
module purge ; module load bioinfo/snakemake-5.25.0 # for test on node
snakemake -np # dry run
sbatch job.sh ; watch 'squeue -u sschmitt' # run
less detMut.*.err # snakemake outputs, use MAJ+F
less detMut.*.out # snakemake outputs, use MAJ+F
snakemake --dag | dot -Tsvg > dag/dag.svg # dag
module purge ; module load bioinfo/snakemake-5.8.1 ; module load system/Python-3.6.3 # for report
snakemake --report report.html # report
```

# Workflow

## Reference

*Copy and index reference for software to work with.*

### [cp_reference](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/cp_reference.smk)

* Tools: `cp`

### [bwa_index](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/bwa_index.smk)

* Tools: [`BWA index`](http://bio-bwa.sourceforge.net/bwa.shtml)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bwa/bwa:latest

### [samtools_faidx](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/samtools_faidx.smk)

* Tools: [`samtools faidx`](http://www.htslib.org/doc/samtools-faidx.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [gatk_dict](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/gatk_dict.smk)

* Tools: [`gatk CreateSequenceDictionary`](https://gatk.broadinstitute.org/hc/en-us/articles/360036729911-CreateSequenceDictionary-Picard-)
* Singularity: docker://broadinstitute/gatk

## Reads

*Copy reads, report quality and trim.*

### [cp_reference](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/cp_reference.smk)

* Tools: `cp`

### [fastqc](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/fastqc.smk)

* Tools: [`fastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/)
* Singularity: docker://biocontainers/fastqc:v0.11.9_cv8

### [multiqc](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/multiqc.smk)

* Tools: [`MultiQC`](https://multiqc.info/)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/multiqc/multiqc:latest

### [trimmomatic](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/trimmomatic.smk)

* Tools: [`Trimmomatic`](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/trimmomatic/trimmomatic:latest

## Alignments

*Align reads against reference and report alignment quality.*

### [bwa_mem](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/bwa_mem.smk)

* Tools: [`BWA mem`](http://bio-bwa.sourceforge.net/bwa.shtml)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bwa/bwa:latest

### [samtools_sort](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/samtools_sort.smk)

* Tools: [`Samtools sort`](http://www.htslib.org/doc/samtools-sort.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [samtools_index](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/samtools_index.smk)

* Tools: [`Samtools index`](http://www.htslib.org/doc/samtools-index.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [gatk_markduplicates](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/gatk_markduplicates.smk)

* Tools: [`gatk MarkDuplicates`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-)
* Singularity: docker://broadinstitute/gatk

### [samtools_stats](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/samtools_stats.smk)

* Tools: [`Samtools stats`](http://www.htslib.org/doc/samtools-stats.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [qualimap](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/qualimap.smk)

* Tools: [`QualiMap`](http://qualimap.conesalab.org/doc_html/command_line.html)
* Singularity: docker://pegi3s/qualimap

## Mutation

*Detect mutation using `freebayes`, `gatk`, and `Mutect2` (`strelka`, `manta`, `tiddit`, etc can be further implemented).*

### Mutect2

#### [gatk_mutect2](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/gatk_mutect2.smk)

* Tools: [`gatk Mutect2`](https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2)
* Singularity: docker://broadinstitute/gatk

<!-- ### freebayes -->

<!-- #### [freebayes](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/freebayes_somatic.smk) -->

<!-- * Tools: [`freebayes`](https://github.com/freebayes/freebayes) -->
<!-- * Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/freebayes/freebayes:latest -->

<!-- ### GATK -->

<!-- #### [gatk_haplotypecaller](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/gatk_haplotypecaller.smk) -->

<!-- * Tools: [`gatk HaplotypeCaller`](https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller) -->
<!-- * Singularity: docker://broadinstitute/gatk -->

<!-- #### [gatk_genotypegvcfs](https://github.com/sylvainschmitt/detectMutations/blob/main/rules/gatk_genotypegvcfs.smk) -->

<!-- * Tools: [`gatk GenotypeGVCFs`](https://gatk.broadinstitute.org/hc/en-us/articles/360037057852-GenotypeGVCFs) -->
<!-- * Singularity: docker://broadinstitute/gatk -->

# Results

```{r }
mutations <- read_tsv("../generateMutations/results/mutations/Qrob_PM1N_Qrob_Chr01_mutated_N100_R2.tsv") %>% 
  mutate(True = 1)
library <- "N100_R2_AF0.8_NR50000"
somatic <- lapply(c(mutect2 = "mutect2"),
                  function(caller) as.data.frame(read.vcfR(file.path("results", library, caller, 
                                                                     paste0(library, ".vcf")), verbose = F)@fix)) %>% 
  bind_rows(.id = "caller") %>% 
  mutate(count = 1) %>% 
  reshape2::dcast(CHROM + POS + REF + ALT ~ caller, value.var = "count") %>% 
  mutate_at(c("CHROM", "REF", "ALT"), as.character) %>% 
  mutate(POS = as.numeric(as.character(POS))) %>% 
  mutate(REF = substr(REF, 1, 1), ALT = substr(ALT, 1, 1))
mutations %>%  
  full_join(somatic) %>% 
  mutate(Caller = "mutect2") %>% # to be changed with dcast with multiple callers 
  rename(Called = mutect2) %>% 
  mutate_at(c("True", "Called"), funs(ifelse(is.na(.), 0, .))) %>% 
  mutate(Confusion = recode(paste0(True, Called), "01" = "False Positive",
                            "10" = "False Negative", "11" = "True Positive")) %>% 
  group_by(Caller, Confusion) %>% 
  summarise(N = n()) %>% 
  group_by(Caller) %>% 
  mutate(P = round(N/sum(N)*100, 2)) %>% 
  kable()
```
