---
title: "detect Mutations - Fruits"
author: Sylvain Schmitt
date: September 29, 2023
output:
  github_document: 
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
rm(list = ls()) ; invisible(gc()) ; set.seed(42)
library(knitr)
library(tidyverse)
opts_chunk$set(echo = F, message = F, warning = F, fig.height = 6, fig.width = 8,
               cache = F, cache.lazy = F)
```


```{r renaming, eval=FALSE}
angela_files <- data.frame(old_file = list.files("data/reads/fruits/")) %>% 
  separate(old_file, c("library", "repetition", "S", "line", 
                       "strand", "one", "file", "compression"), remove = FALSE) %>% 
  select(old_file, library, strand) %>% 
  left_join(readxl::read_xlsx("data/treemutation_fruits.xlsx", "Angela") %>% 
              mutate(library = paste0(Code_Espece, Tube))) %>% 
  mutate(new_file = paste0(gsub(".", "_", Echantillon, fixed = TRUE),
                           "_", strand, ".fq.gz"))
angela_files %>% 
  mutate(command = paste("mv", old_file, new_file)) %>% 
  select(command) %>% 
  write_tsv("data/reads/fruits/command.sh")
mutate(angela_files, library = paste0('"', gsub(".", "_", Echantillon, fixed = TRUE), '"'))$library %>% 
  unique() %>% 
  paste0(collapse = ", ") %>% 
  cat()

sixto_files <- data.frame(old_file = list.files("data/reads/sixto/")) %>% 
  separate(old_file, c("library", "repetition", "S", "line", 
                       "strand", "one", "file", "compression"), remove = FALSE) %>% 
  select(old_file, library, strand) %>% 
  left_join(readxl::read_xlsx("data/treemutation_fruits.xlsx", "Sixto") %>% 
              mutate(library = paste0(Code_Espece, Tube))) %>% 
  mutate(new_file = paste0(gsub(".", "_", Echantillon, fixed = TRUE),
                           "_", strand, ".fq.gz"))
sixto_files %>% 
  mutate(command = paste("mv", old_file, new_file)) %>% 
  select(command) %>% 
  write_tsv("data/reads/sixto/command.sh")
mutate(sixto_files, library = paste0('"', gsub(".", "_", Echantillon, fixed = TRUE), '"'))$library %>% 
  unique() %>% 
  paste0(collapse = ", ") %>% 
  cat()
```

```{r bedfiles, eval=FALSE}
read_tsv("data/angela_fruits_candidate_mutations.tsv") %>% 
  mutate(START = POS, END = POS) %>% 
  select(CHROM, START, END) %>% 
  write_tsv("data/angela.bed", col_names = FALSE)
read_tsv("data/sixto_fruits_candidate_mutations.tsv") %>% 
  mutate(START = POS-1, END = POS+1) %>% 
  select(CHROM, START, END) %>% 
  write_tsv("data/sixto.bed", col_names = FALSE)
```

[`singularity` & `snakemake`](https://github.com/sylvainschmitt/snakemake_singularity) workflow to detect mutations in fruits from Angela and Sixto.


# Installation

- [x] Python ≥3.5
- [x] Snakemake ≥5.24.1
- [x] Golang ≥1.15.2
- [x] Singularity ≥3.7.3
- [x] This workflow

```{bash, eval=F, echo=T}
# Python
sudo apt-get install python3.5
# Snakemake
sudo apt install snakemake`
# Golang
export VERSION=1.15.8 OS=linux ARCH=amd64  # change this as you need
wget -O /tmp/go${VERSION}.${OS}-${ARCH}.tar.gz https://dl.google.com/go/go${VERSION}.${OS}-${ARCH}.tar.gz && \
sudo tar -C /usr/local -xzf /tmp/go${VERSION}.${OS}-${ARCH}.tar.gz
echo 'export GOPATH=${HOME}/go' >> ~/.bashrc && \
echo 'export PATH=/usr/local/go/bin:${PATH}:${GOPATH}/bin' >> ~/.bashrc && \
source ~/.bashrc
# Singularity
mkdir -p ${GOPATH}/src/github.com/sylabs && \
  cd ${GOPATH}/src/github.com/sylabs && \
  git clone https://github.com/sylabs/singularity.git && \
  cd singularity
git checkout v3.7.3
cd ${GOPATH}/src/github.com/sylabs/singularity && \
  ./mconfig && \
  cd ./builddir && \
  make && \
  sudo make install
# detect Mutations
git clone git@github.com:sylvainschmitt/detectMutations.git
cd detectMutations
git checkout fruits
```

# Usage

## Locally

```{bash, eval=F, echo=T}
snakemake -np -j 3 --resources mem_mb=10000 # dry run
snakemake --dag | dot -Tsvg > dag/dag.svg # dag
snakemake --use-singularity -j 3 --resources mem_mb=10000 # run
```

## HPC

```{bash, eval=F, echo=T}
module load bioinfo/snakemake-5.25.0 # for test on node
snakemake -np # dry run
sbatch job.sh # run
snakemake --dag | dot -Tsvg > dag/dag.svg # dag
```

# Workflow

## Reference

*Copy and index reference.*

### [cp_reference](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/cp_reference.py)

  - Tools: `cp`

### [bwa_index](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/bwa_index.py)

  - Tools: [`BWA index`](http://bio-bwa.sourceforge.net/bwa.shtml)
  - Singularity:
    oras://registry.forgemia.inra.fr/gafl/singularity/bwa/bwa:latest
    
### [samtools_faidx](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/samtools_faidx.py)

  - Tools: [`samtools faidx`](http://www.htslib.org/doc/samtools-faidx.html)
  - Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [gatk_dict](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/gatk_dict.py)

  - Tools: [`gatk CreateSequenceDictionary`](https://gatk.broadinstitute.org/hc/en-us/articles/360037422891-CreateSequenceDictionary-Picard-)
  - Singularity: docker://broadinstitute/gatk

## Reads

*Trim reads.*

### [trimmomatic](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/trimmomatic.py)

* Tools: [`Trimmomatic`](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/trimmomatic/trimmomatic:latest

## Alignments

*Align reads against reference, mark duplicated, and report alignment quality.*

### [bwa_mem](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/bwa_mem.py)

* Tools: [`BWA mem`](http://bio-bwa.sourceforge.net/bwa.shtml)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/bwa/bwa:latest

### [samtools_view](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/samtools_view.py)

* Tools: [`Samtools view`](http://www.htslib.org/doc/samtools-view.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [samtools_sort](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/samtools_sort.py)

* Tools: [`Samtools sort`](http://www.htslib.org/doc/samtools-sort.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

### [samtools_index](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/samtools_index.py)

* Tools: [`Samtools index`](http://www.htslib.org/doc/samtools-index.html)
* Singularity: oras://registry.forgemia.inra.fr/gafl/singularity/samtools/samtools:latest

## Variants

*Detect mutations.*

### [gatk_haplotypecaller](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/gatk_haplotypecaller.py)

* Tools: [`gatk HaplotypeCaller`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-)
* Singularity: docker://broadinstitute/gatk

### [gatk_genomicdbimport](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/gatk_genomicdbimport.py)

* Tools: [`gatk GenomicsDBImport`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-)
* Singularity: docker://broadinstitute/gatk

### [gatk_genotypegvcfs](https://github.com/sylvainschmitt/detectMutations/blob/fruits/rules/ggatk_genotypegvcfs.py)

* Tools: [`gatk GenomicsDBImport`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-)
* Singularity: docker://broadinstitute/gatk

# Results

```{r angelares, eval=FALSE}
library(vcfR)
angela <- read.vcfR("results/angela.vcf")
raw <- angela@fix %>% 
  as_tibble() %>% 
  separate_rows(INFO, sep = ";") %>% 
  separate(INFO, c("metric", "value"), convert = TRUE, sep = "=") %>% 
  pivot_wider(names_from = metric, values_from = value) %>% 
  bind_cols(angela@gt %>% 
              as_tibble())
candidates <- read_tsv("data/angela_fruits_candidate_mutations.tsv")
joined <- raw %>% 
  filter(paste(CHROM, POS, REF, ALT) %in% 
           with(candidates, paste(CHROM, POS, REF, ALT))) %>% 
  select(-ID, -FILTER, -AN, -ExcessHet, -FS, -InbreedingCoeff, -MLEAC, -MLEAF,
         -MQ, -QD, -SOR, -BaseQRankSum, -MQRankSum, -ReadPosRankSum, -AC) %>% 
  gather(library, values, -CHROM, -POS, -REF, -ALT, -QUAL, -AF, -DP, -FORMAT) %>% 
  mutate(POS = as.numeric(POS)) %>% 
  left_join(candidates) %>% 
  separate(values, c("GT", "AD", "DP_lib"),":", convert = TRUE) %>% 
  select(-FORMAT) %>% 
  separate(AD, c("AC_REF", "AC_ALT"), convert = TRUE) %>% 
  mutate(AF_lib = AC_ALT / DP_lib)
filtered <- joined %>% 
  filter(AF_lib > 0) %>% 
  separate(library, c("species", "branch_mutation", "fruit", "tissue"), 
           remove = FALSE) %>% 
  separate_rows(branch) %>% 
  filter(branch_mutation == branch)
ggplot(filtered, aes(DP_lib, fill = (DP_lib > 9))) +
  geom_histogram() +
  theme_bw() +
  xlab("Sequencing depth") + ylab("# of library") +
  ggtitle("Angela") +
  scale_fill_discrete(guide = "none")
ggplot(filtered, aes(AF_lib, fill = tissue)) +
  geom_histogram() +
  theme_bw() +
  facet_wrap(~ gsub("Super-Scaffold_", "", SNV), scales = "free_y") +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Angela AF in fruits", "All depth") + xlab("")
ggplot(filter(filtered, DP_lib > 4),
       aes(AF_lib, fill = tissue)) +
  geom_histogram() +
  theme_bw() +
  facet_wrap(~ gsub("Super-Scaffold_", "", SNV), scales = "free") +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Angela AF in fruits", "Depth above 9 reads") + xlab("") + xlim(0,1)
```

```{r sixtores, eval=FALSE}
library(vcfR)
sixto <- read.vcfR("results/sixto.vcf")
raw <- sixto@fix %>% 
  as_tibble() %>% 
  separate_rows(INFO, sep = ";") %>% 
  separate(INFO, c("metric", "value"), convert = TRUE, sep = "=") %>% 
  pivot_wider(names_from = metric, values_from = value) %>% 
  bind_cols(sixto@gt %>% 
              as_tibble())
candidates <- read_tsv("data/sixto_fruits_candidate_mutations.tsv")
joined <- raw %>% 
  filter(paste(CHROM, POS, REF, ALT) %in% 
           with(candidates, paste(CHROM, POS, REF, ALT))) %>% 
  select(-ID, -FILTER, -AN, -ExcessHet, -FS, -InbreedingCoeff, -MLEAC, -MLEAF,
         -MQ, -QD, -SOR, -BaseQRankSum, -MQRankSum, -ReadPosRankSum, -AC) %>% 
  gather(library, values, -CHROM, -POS, -REF, -ALT, -QUAL, -AF, -DP, -FORMAT) %>% 
  mutate(POS = as.numeric(POS)) %>% 
  left_join(candidates) %>% 
  separate(values, c("GT", "AD", "DP_lib"),":", convert = TRUE) %>% 
  select(-FORMAT) %>% 
  separate(AD, c("AC_REF", "AC_ALT"), convert = TRUE) %>% 
  mutate(AF_lib = AC_ALT / DP_lib)
filtered <- joined %>% 
  filter(AF_lib > 0) %>% 
  separate(library, c("species", "branch_mutation", "fruit", "tissue"), 
           remove = FALSE) %>% 
  separate_rows(branch) %>% 
  filter(branch_mutation == branch)
ggplot(filtered, aes(DP_lib, fill = (DP_lib > 9))) +
  geom_histogram() +
  theme_bw() +
  xlab("Sequencing depth") + ylab("# of library") +
  ggtitle("Sixto") +
  scale_fill_discrete(guide = "none")
ggplot(filtered, aes(AF_lib, fill = tissue)) +
  geom_histogram() +
  theme_bw() +
  facet_wrap(~ gsub("Super-Scaffold_", "", SNV), scales = "free_y") +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Sixto AF in fruits", "All depth") + xlab("")
ggplot(filter(filtered, DP_lib > 4),
       aes(AF_lib, fill = tissue)) +
  geom_histogram() +
  theme_bw() +
  facet_wrap(~ gsub("Super-Scaffold_", "", SNV), scales = "free_y") +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Sixto AF in fruits", "Depth above 9 reads") + xlab("") + xlim(0,1)
```


